x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -195,42 +193,29 @@ private void ensureFileRealmEnabled(Settings settings) throws Exception {
     private void checkClusterHealthWithRetries(Environment env, Terminal terminal, String username, SecureString password, int retries,
                                                boolean force) throws Exception {
         CommandLineHttpClient client = clientFunction.apply(env);
-        final URL clusterHealthUrl = createURL(new URL(client.getDefaultURL()), "_cluster/health", "?pretty");
-        final HttpResponse response;
         try {
-            response = client.execute("GET", clusterHealthUrl, username, password, () -> null, this::responseBuilder);
+            client.checkClusterHealthWithRetriesWaitingForCluster(username, password, 0, force);
         } catch (Exception e) {
-            throw new UserException(ExitCodes.UNAVAILABLE, "Failed to determine the health of the cluster. ", e);
-        }
-        final int responseStatus = response.getHttpStatus();
-        if (responseStatus != HttpURLConnection.HTTP_OK) {
-            // We try to write the roles file first and then the users one, but theoretically we could have loaded the users
-            // before we have actually loaded the roles so we also retry on 403 ( temp user is found but has no roles )
-            if ((responseStatus == HttpURLConnection.HTTP_UNAUTHORIZED || responseStatus == HttpURLConnection.HTTP_FORBIDDEN)
-                && retries > 0 ) {
-                terminal.println(
-                    Terminal.Verbosity.VERBOSE,
-                    "Unexpected http status [" + responseStatus + "] while attempting to determine cluster health. Will retry at most "
-                        + retries
-                        + " more times."
-                );
-                Thread.sleep(1000);
-                retries -= 1;
-                checkClusterHealthWithRetries(env, terminal, username, password, retries, force);
+            if (e instanceof ElasticsearchStatusException &&
```

Comment:
This doesn't work now, because `checkClusterHealthWithRetriesWaitingForCluster` wraps exception in `IllegalStateException`s.

I'm starting to regret my suggestion to have a common approach for the cluster health polling.
My suggestion now is to do whatever works that follows the following requirements:

For `BaseRunAsSuperUserCommand`:
* fail on connection exception (node is down)
* retry on HTTP_UNAUTHORIZED and HTTP_FORBIDDEN

For `BootstrapPasswordAndEnrollmentTokenForInitialNode`:
* retry on connection exception
* fail on any Exception

I think it's better that these methods be at the `CommandLineHttpClient` level, to aid with discoverability, and that they reuse code, but I'm not going to gripe about it.

Reply:
`checkClusterHealthWithRetriesWaitingForCluster` is in `CommandLineHttpClient`
`checkClusterHealthWithRetries` is in `BaseRunAsSuperuserCommand` because it is superuser aware command

`checkClusterHealthWithRetriesWaitingForCluster` throws `ElasticsearchStatusException` in case of HTTP_UNAUTHORIZED and HTTP_FORBIDDEN and pass the specific http error code to `checkClusterHealthWithRetries` so it can retry in this case

Reply:
> I'm starting to regret my suggestion to have a common approach for the cluster health polling.

++ :/ I think we gave it a try and it doesn't work ( I don't like exception type based flow control ) . I'd suggest we leave the BaseRunAsSuperUserCommand check health method as it was and implement  something for BootstrapPasswordAndEnrollmentTokenForInitialNode that makes sense for it since these two have slightly different requirements. I'd leave the first in the BaseRunAsSuperUserCommand as it is very specific to it, but I don't have strong opinions on where to place the second checkclusterhealth method. 

WDYT Lyudmila ?

Reply:
I tend to agree with both of you. I think the current implementation would work, but it doesn't look intuitive and overall fragil.
I would leave the one waiting for cluster in `CommandLineHttpClient` as I think it is generic enough to be used by other commands